File: packages/tonic-core/src/hydration/repoStructure.ts
import * as fs from "node:fs";
import * as path from "node:path";

export type RepoStructureArtifactV1 = {
  schema: "tonic-repo-structure";
  version: "1";
  repo_root: string;
  top_level: string[];
  max_depth: number;
};

export function summarizeRepoStructure(repoRoot: string, maxDepth = 2): RepoStructureArtifactV1 {
  const top: string[] = [];
  try {
    for (const e of fs.readdirSync(repoRoot, { withFileTypes: true })) {
      if (e.name === ".git" || e.name === "node_modules" || e.name === ".tonic") {
        continue;
      }
      top.push(e.name);
    }
  } catch {
    /* ignore */
  }
  top.sort();
  return {
    schema: "tonic-repo-structure",
    version: "1",
    repo_root: repoRoot.replace(/\\/g, "/"),
    top_level: top,
    max_depth: maxDepth,
  };
}

export function writeRepoStructure(pathOut: string, art: RepoStructureArtifactV1): void {
  fs.mkdirSync(path.dirname(path.resolve(pathOut)), { recursive: true });
  fs.writeFileSync(pathOut, JSON.stringify(art, null, 2) + "\n", "utf8");
}

export function repoStructureExcerpt(art: RepoStructureArtifactV1): string {
  return `Top-level: ${art.top_level.join(", ")}`;
}

File: merge-tonic-lib/tonic/hydration/repo_structure.py
"""Repo structure summary (parity with repoStructure.ts)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def summarize_repo_structure(repo_root: str, max_depth: int = 2) -> dict[str, Any]:
    top: list[str] = []
    try:
        for e in sorted(Path(repo_root).iterdir(), key=lambda x: x.name):
            if e.name in (".git", "node_modules", ".tonic"):
                continue
            top.append(e.name)
    except OSError:
        pass
    return {
        "schema": "tonic-repo-structure",
        "version": "1",
        "repo_root": str(Path(repo_root).resolve()).replace("\\", "/"),
        "top_level": top,
        "max_depth": max_depth,
    }


def repo_structure_excerpt(art: dict[str, Any]) -> str:
    tl = art.get("top_level") or []
    if isinstance(tl, list):
        return "Top-level: " + ", ".join(str(x) for x in tl)
    return "Top-level: "


def write_repo_structure(path_out: str, art: dict[str, Any]) -> None:
    p = Path(path_out).resolve()
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(json.dumps(art, indent=2) + "\n", encoding="utf-8")
